src/dynamic.cpp

```diff
@@ -500,3 +460,174 @@ void DYNAMIC::Search::print_result() const {
 
   std::cout << " nodes " << (totalCounter + counter);
 }
+
+namespace {
+
+    // Check if the position is semistatically unwinnable with recursive trivial progress.
+    bool is_unwinnable_with_trivial_progress(Position& pos, Color intendedWinner) {
+        MoveList<LEGAL> moveList(pos);
+
+        // Checkmate or Stalemate
+        if (moveList.size() == 0)
+            return !pos.checkers() || pos.side_to_move() == intendedWinner;
+
+        // Recursive trivial progress
+        if (moveList.size() == 1)
+        {
+            StateInfo stateInfo;
+            pos.do_move(*moveList.begin(), stateInfo);
+
+            // Returns true if the position is repeated.
+            bool res = stateInfo.repetition || is_unwinnable_with_trivial_progress(pos, intendedWinner);
+
+            pos.undo_move(*moveList.begin());
+            return res;
+        }
+
+        return SemiStatic::is_unwinnable(pos, intendedWinner);
+    }
+
+    bool side_to_move_can_capture_king(const Position& pos) {
+
+        for (const auto& m : MoveList<LEGAL>(pos)) {
+            if (!pos.empty(to_sq(m)) && type_of(pos.piece_on(to_sq(m))) == KING)
+                return true;
+        }
+        return false;
+    }
+
+    // Apply iterative deepening (find_mate may look deeper than maxDepth on
+    // rewarded variations)
+    DYNAMIC::SearchResult iterative_deepening(Position& pos, DYNAMIC::Search& search) {
+        for (int maxDepth = 2; maxDepth <= 1000; maxDepth++) {
+            // This choice seems empirically good
+            uint64_t limit = 10000;
+            search.set(maxDepth, search.actual_depth(), limit);
+            bool mate =
+                find_mate<DYNAMIC::FULL, DYNAMIC::ANY>(pos, search, 0, false, false);
+
+            if (!search.is_interrupted() && !mate)
+                search.set_unwinnable();
+
+            if (search.get_result() != DYNAMIC::UNDETERMINED ||
+                search.is_limit_reached())
+                break;
+        }
+
+        return search.get_result();
+    }
+}
+
+DYNAMIC::SearchResult DYNAMIC::full_analysis(Position& pos, DYNAMIC::Search& search) {
+    search.init();
+    search.set(0, 0, 0);
+
+    if (side_to_move_can_capture_king(pos)) {
+        search.set_unwinnable(); // is it Ok?
+        return search.get_result();
+    }
+
+    // Required to detect repetitions 
+    assert(pos.state()->pliesFromNull == 0);
+
+    std::deque<StateInfo> states;
+
+    // Trivial progress
+    while (true) {
+        MoveList<LEGAL> moveList(pos);
+
+        if (moveList.size() == 1) {
+            states.push_back(StateInfo());
+            pos.do_move(*moveList.begin(), states.back());
+            search.annotate_move(*moveList.begin());
+            search.step();
+
+            // If a position is forced to repeat, then it is unwinnable.
+            if (pos.state()->repetition) {
```

Comment:
How does this work? Does it detect repetitions that occurred many moves ago?

Reply:
Yes. Position::do_move() calculates, among other things, repetition info:
```
  // Calculate the repetition info. It is the ply distance from the previous
  // occurrence of the same position, negative in the 3-fold case, or zero
  // if the position was not repeated.
  st->repetition = 0;
  int end = std::min(st->rule50, st->pliesFromNull);
  if (end >= 4)
  {
      StateInfo* stp = st->previous->previous;
      for (int i = 4; i <= end; i += 2)
      {
          stp = stp->previous->previous;
          if (stp->key == st->key)
          {
              st->repetition = stp->repetition ? -i : i;
              break;
          }
      }
  }
```